### PR TITLE
ExternalIntegrationLinks table

### PR DESCRIPTION
**Pull request by EdwinGuzman**
First part to resolve [SIMPLY-2296](https://jira.nypl.org/browse/SIMPLY-2296). Updates are mostly in mirror.py to remove the sitewide configuration and allow for multiple mirrors for different purposes.

### CODE REVIEW

File: s3.py
Comment (EdwinGuzman) [inline s3.py]: If I understand this correctly, under "Storage" a purpose for that storage integration is assigned.
Then, in the External Catalogs, you'd only see storages that should be used for MARC files. Similar for collections for covers and books storages.

The externalintegrationlink table would have something where the MARC exporter integration has a storage integration as its linked integration and the purpose is "MARC"?

File: marc.py
Comment (EdwinGuzman) [inline marc.py]: Should the list of s3 integrations for MARC files be generated here? I'm currently setting it up but in circulation.

Comment (leonardr) [reply marc.py]: You could have a method `ExternalIntegration.storage_for(self, _db, artifact_type)`which found all of the storage integrations that were suitable for storing certain things. "Suitability" would be hard-coded -- we know that an S3 integration can be used for anything so long as you've defined an appropriate bucket for that purpose, and we know that an LCP integration can only be used to store ebooks. I think this code could definitely go into core.

File: marc.py
Comment (leonardr) [inline marc.py]: It looks like `integration` is required, and if so, you shouldn't set a default of `None` in the method signature.

File: model/configuration.py
Comment (leonardr) [inline model/configuration.py]: These are the purposes, right?

File: model/configuration.py
Comment (leonardr) [inline model/configuration.py]: I think index `purpose` as well, just to be safe.

File: migration/20190924-add-externalintegrationslink-table.sql
Comment (leonardr) [inline migration/20190924-add-externalintegrationslink-table.sql]: I tried this out on my laptop and it looks like this doesn't create the indexes. However, when you're creating a brand new database table you don't need a migration script at all. The database table will be created automatically the first time SQLAlchemy is initialized.

File: opds_import.py
Comment (leonardr) [inline opds_import.py]: This is a great payoff -- it's totally clear what's happening here.

However, it's also clear that `mirror` now is just for the book covers. We're going to need a separate `MirrorUploader` to store any ebooks we find. In this code, instead of self.mirror we should be using self.cover_mirror or self.book_mirror.

File: scripts.py
Comment (leonardr) [inline scripts.py]: Again, we now have different MirrorUploaders for different types of items. So  `ReplacementPolicy` needs to take two mirrors -- one for every type of thing that might be mirrored. It might be better to make `ReplacementPolicy` take a dictionary that maps "purpose" to an appropriate MirrorUploader.

File: model/configuration.py
Comment (EdwinGuzman) [reply model/configuration.py]: Yes.

File: migration/20190924-add-externalintegrationslink-table.sql
Comment (EdwinGuzman) [reply migration/20190924-add-externalintegrationslink-table.sql]: I see. So then this isn't really needed at all?

Comment (leonardr) [reply migration/20190924-add-externalintegrationslink-table.sql]: Yeah, you can lose it.

File: opds_import.py
Comment (leonardr) [reply opds_import.py]: Or we should have a dictionary that maps a purpose to a MirrorUploader designed for that purpose.

File: marc.py
Comment (EdwinGuzman) [reply marc.py]: That makes sense but I'm giving the type of bucket each integration is used for when it is added to a collection or catalog service (it's "purpose"). So here, it wouldn't be able to know that some buckets are specifically for, say, MARC files. In circulation, I was listing all the available buckets regardless of how they could be used.

Comment (leonardr) [reply marc.py]: That's fine -- I'm only suggesting you be able to grab the buckets that _could_ be used for MARC files because it's technically possible.

File: scripts.py
Comment (leonardr) [inline scripts.py]: I'm pretty sure this is stylistic, but i would write `if covers or books:` and build the dictionary inside -- otherwise you're making a dictionary you're not using.

File: marc.py
Comment (leonardr) [inline marc.py]: I'm at the very beginning of this diff so maybe this will make sense later, but how do the mirror integrations get added to `options`?

I would expect the config here to look more like:

```
"type": "storage_integration",
"storage_purpose": "MARC",
```

File: marc.py
Comment (leonardr) [inline marc.py]: I see, this method gets called at some point and it adds all the existing storage integrations to the list.

This is fine but--again, just thinking out loud as I go through the diff--it seems like you're going to need to have similar code for  setting up the "cover" and "book" ExternalIntegrationLinks for a collection.

This can wait, but the other issue is that not all ExternalIntegrations with STORAGE_GOAL can actually store MARC records. An S3 integration needs to have a MARC bucket defined, and an LCP integration won't be able to store MARC records at all.

File: marc.py
Comment (leonardr) [reply marc.py]: This is looking good, but `integration` is too vague a term. How about `mirror_integration`?

File: metadata_layer.py
Comment (leonardr) [inline metadata_layer.py]: You need to handle the case where `mirror_type` doesn't show up in the dictionary.

File: metadata_layer.py
Comment (leonardr) [inline metadata_layer.py]: I don't understand this piece. I feel like this is the point where you need to decide which of the mirrors in replace.mirror to use for this particular link -- so you do need new code here and looking at replace.mirror is appropriate. But I would expect it to have something to do with `link.rel` -- images go to the image mirror, books go to the book mirror. Maybe there needs to be a `ReplacementPolicy.mirror_for(LinkData)` method to sort this out.

I think the code you have is going to tell _every_ mirror to mirror every link.

File: metadata_layer.py
Comment (leonardr) [inline metadata_layer.py]: Again, I think every mirror is going to mirror every link. And this looks like duplicate code (though it was duplicate code before you changed it). Maybe the right thing to do is to put all the code in `mirror_link`, and make `mirror_link` figure out whether the link should actually be mirrored, and if so, which is the right mirror to use.

File: mirror.py
Comment (leonardr) [inline mirror.py]: It's good to have a docstring here but we're doing this to eventually support MirrorUploaders other than S3, so let's not start with the assumption that this is an S3 integration specifically.

File: model/configuration.py
Comment (leonardr) [inline model/configuration.py]: For consistency with other methods this should be called `get_integrations_for_goal`.

File: model/configuration.py
Comment (leonardr) [inline model/configuration.py]: You don't need to call `all()` here -- the query wil lbe run when the results are needed. This will allow someone to call `ExternalIntegration.for_goal` and then tack on any extra query conditions they need.

File: model/configuration.py
Comment (leonardr) [inline model/configuration.py]: This error message should also mention the collection name.

File: opds_import.py
Comment (leonardr) [inline opds_import.py]: In ReplacementPolicy you have a dictionary of mirrors keyed by purpose, and I think it's best to do the same here. Not only will this make everything consistent, this will let us add other kinds of mirrors (such as for book previews) without changing the method signature.

File: opds_import.py
Comment (leonardr) [inline opds_import.py]: Yeah, you end up with a dictionary, so take a dictionary as input and make your own if the input is `None`.

File: s3.py
Comment (leonardr) [inline s3.py]: I've been saying "book bucket" when I actually need the "open_access_content_bucket". We shouldn't need a separate "book bucket" because we'll probably never store non-open-access content on S3.

We do need a 'books' _purpose_ (which indicates that an S3 integration can be used to store open-access books or an LCP integration can be used to store any kind of books), but that's different and you already have that.

File: marc.py
Comment (leonardr) [inline marc.py]: This works but you don't need to call `all`. Iterating over the query object will run the query.

File: metadata_layer.py
Comment (EdwinGuzman) [reply metadata_layer.py]: So would it make sense to assume that the link should be mirrored for cover images if the `rel` is in `MIRRORED = [OPEN_ACCESS_DOWNLOAD, IMAGE, THUMBNAIL_IMAGE]`? Otherwise, it is a book mirror. In this specific case, there is a check to see if `rel` is in `CIRCULATION_ALLOWED` so that limits what type it is and I guess in this case, it is a book mirror?

Comment (leonardr) [reply metadata_layer.py]: Yes, it sounds like in this case it's a book mirror. For `[IMAGE, THUMBNAIL_IMAGE]` it would be a cover mirror.

File: metadata_layer.py
Comment (leonardr) [inline metadata_layer.py]: This doesn't use `self` at all. As is, this would be a better fit in `LinkData`. It would make sense as part of `ReplacementPolicy` if it looked in `self.mirrors` and returned the actual mirror object to use.

Normally I don't think this is a big deal but `metadata_layer.py`, like the model, is a semi-public API that's used by a lot of different code so I'm more concerned with the API design.

File: metadata_layer.py
Comment (leonardr) [inline metadata_layer.py]: However you resolve the `mirror_type_for` issue, it's now possible to put _all_ of this logic inside `mirror_link`. You're already passing the LinkData and the ReplacementPolicy into that method, and that gives it all the information necessary to figure out whether, how and where to do the mirroring.

**Review by leonardr — COMMENTED** [review]
This is looking good. I think you can make the metadata-layer "mirror this link" code easier to use and then we'll be ready.

**Review by leonardr — APPROVED** [review]
:sunglasses:

File: model/configuration.py
Comment (EdwinGuzman) [inline model/configuration.py]: Is this what you meant for https://github.com/NYPL-Simplified/circulation/pull/1333/files#r333120505 ?

Comment (leonardr) [reply model/configuration.py]: Yes, but `_get_settings` is a generic name and this doesn't need to be a function. You can make a list of COLLECTION_MIRROR_SETTINGS and then use it from controller code.

**Comment by leonardr** [comment]
This is looking good overall. We're missing two things:

1. Migration script that drops collection.mirror_integration_id and (if it's set) replaces it with two ExternalIntegrationLinks.
2. Make it possible to set the data source for a "Manual import" integration from the admin interface.

#2 can wait for later but I would like to see #1 in the core branch before merging.

File: model/configuration.py
Comment (EdwinGuzman) [inline model/configuration.py]: @leonardr this worked great, thanks!

Comment (leonardr) [reply model/configuration.py]: Great. Hold off merging this until the circulation branch is ready to go, but I think this is done once you add one more test.

File: tests/models/test_configuration.py
Comment (leonardr) [inline tests/models/test_configuration.py]: In general you can't rely on objects coming out of a database query in a specific order. The simplest way to enforce this is to add `.order_by(ExternalIntegrationLink.other_integration_id)` to the `qu`. You can also do this by comparing two `set`s.

**Comment by leonardr** [comment]
I'd like @aslagle to double-check the migration script (https://github.com/NYPL-Simplified/server_core/pull/1121/files#diff-d9b951efad2ec7b38b0ad144e4df994e), both for its functionality and for a problem I think I see in it:

This migration script uses a table that is only created when the SQLAlchemy model is loaded. DatabaseMigrationScript tries really hard not to load the SQLAlchemy model. I know I said earlier that you didn't need to create the table inside the migration script, and that's generally true, because generally the new table starts out empty.

This situation is unusual because we need the table to start out with a couple of rows. The only time we can add these rows is during a migration script, and (by design) the table won't get added until after the migration script runs. So I think you need to bring back the SQL you had earlier that creates the database table with all of its indexes and foreign keys. Running `pg_dump --schema-only --schema=externalintegrationlinks` on a working database should give you the SQL to use.

**Review by leonardr — CHANGES_REQUESTED** [review]
Un-approving because of concerns about the existing migration script.

**Comment by aslagle** [comment]
Yeah, I agree. The table needs to be created as part of the migration in this case. Also, seems like the purpose should be more specific, like "mirror_books" and "mirror_covers".

**Comment by leonardr** [comment]
I'm okay with adding "mirror_" or "storage_" to the purpose of links used for storage.

File: migration/20191010-update-external-integration-mirrors.sql
Comment (leonardr) [inline migration/20191010-update-external-integration-mirrors.sql]: Still not quite right -- this is inside the first BEGIN statement. I ran it on my dev database and got the warning as expected, but then an error, "multiple primary keys for table "externalintegrationslinks" are not allowed", which would have caused the script to fail.

**Review by leonardr — APPROVED** [review]
Looks good now!

**Pull request merged**
